zerver/lib/test_runner.py

```diff
@@ -463,6 +441,39 @@ def run_tests(self, test_labels, extra_tests=None,
             print()
             sys.exit(1)
 
+        # Checking for import failures in serial mode
+        if self.parallel == 1 and hasattr(suite, '_tests') and len(suite._tests):
```

Comment:
I think you can just loop over the `suite` and check for the import failure in the name of the test:
```
for test in suite:
   if test_name.startswith(import_failure_prefix_old):
      # Code
   elif test_name.startswith(import_failure_prefix_new):
      # Code
```

Reply:
That would be great :) But `suite` here is either an instance of `TestSuite` or `ParallelTestSuite`, and neither of them returns test names when being iterated over.

Reply:
You can use this:

```
diff --git a/zerver/lib/test_runner.py b/zerver/lib/test_runner.py
index 29a1daa..5011b48 100644
--- a/zerver/lib/test_runner.py
+++ b/zerver/lib/test_runner.py
@@ -442,11 +442,11 @@ class Runner(DiscoverRunner):
             sys.exit(1)
                                                                                                                                                                                                                   
         # Checking for import failures in serial mode
-        if self.parallel == 1 and hasattr(suite, '_tests') and len(suite._tests):
-            if hasattr(suite._tests[0], '_exception'):
-                print()
-                print(suite._tests[0]._exception)
-                sys.exit(1)
+        if isinstance(suite, TestSuite):
+            for test in suite:
+                test_name = full_test_name(test)
+                if 'ModuleImportFailure' in test_name:
+                    print(test_name)
 
         # Checking for import failures in parallel mode
         elif hasattr(suite, 'subsuites'):
```

Reply:
Unfortunately, I don't think I can use this.

`print_error_message` is supposed to be called with a test name of the following form:
```
unittest.loader._FailedTest.zerver.tests.test_audit_log
```
Calling `full_test_name` here gives us something like this:
```
unittest.loader._FailedTest.test_audit_log
```
so `print_error_message` wouldn't work here:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module named 'test_audit_log'
```